src/main/java/com/pusher/client/channel/impl/ChannelManager.java

```diff
@@ -89,6 +98,12 @@ public void unsubscribeFrom(final String channelName) {
             return;
         }
         if (connection.getState() == ConnectionState.CONNECTED) {
+
+            // only private encrypted channels need to be securely torn down
+            if (channel instanceof PrivateEncryptedChannelImpl) {
+                ((PrivateEncryptedChannelImpl) channel).tearDownChannel();
```

Comment:
This class shouldn't make a dependency on specific implementation details. That could be handled within the existing `updateState`. Also I'm not sure if it's tread-safe or generally good to allow the caller's thread access and modify the channel object... (it seems that everywhere there's a barrier of the thread executor queue).

Reply:
this has been refactored and is no longer an issue.